### PR TITLE
Feature: Outfit tags

### PR DESCRIPTION
**Pull request by RisingLeaf**
## Feature Details
Adds basically three things:
in an outfits definition:
`tags <tagname>`
a game action that removes all outfits with the tag on the flagship:
```
"modify ship" [fleet, flagship, parked, system]
    "take outfits"
        [not] tags [list of tags]
        [not] attributes [list of attributes]
```
and a condition that counts all outfits with the tag on the flagship:
`"outfit tag: <tagname>" </>/== count`

Some examples on what it could be used:
It originally came up during a discussion for different hyperdrives for different species, what would not work because people sometimes have to remove hyperdrives during missions. with this tag system, you could group hyperdrives under a tag and remove all of them at once if you want to.

## Testing Done
Not yet tested

## Performance Impact
N/A

### CODE REVIEW

**Comment by petervdmeer** [comment]
I'm missing some background information on why we would want to add tag-functionality to statically defined outfits. What problems do those tags solve that are currently not solved yet? (And do we have an issue where alternatives were discussed?)

**Comment by RisingLeaf** [comment]
> I'm missing some background information on why we would want to add tag-functionality to statically defined outfits. What problems do those tags solve that are currently not solved yet? (And do we have an issue where alternatives were discussed?)

One of the most important thing is different hyperdrives for different factions. With tags you can just group all hyperdrives under one tag and if you need to remove them, what happens in game, you can just use the tag

**Comment by RisingLeaf** [comment]
Or you could remove all cooling, things like that

**Comment by warp-core** [comment]
I'm not sure this is actually the main impediment to adding different hyperdrives.

As I think I've mentioned on Discord, if a different faction actually has a hyperdrive unique enough to merit being its own outfit, perhaps the Syndicate engineers shouldn't be taking it from me anyway? Perhaps they just shouldn't be able to identify it.

"As the Syndicate employees examine your ship, they become progressively more perplexed. It seems they cannot identify what your ship is using as a hyperdrive, and they seem too embarrassed at being unable to find the outfit to listen to your directions. They end up just installing the jump drive in addition to what you alread to have."

If the hyperdrive is actually unique, they definitely shouldn't take it.

And, removing every outfit with the given tag doesn't seem like good behaviour either.

**Comment by petervdmeer** [comment]
> One of the most important thing is different hyperdrives for different factions. With tags you can just group all hyperdrives under one tag and if you need to remove them, what happens in game, you can just use the tag

We could also support this without additional tags, if we add support for removing Outfits based on Outfit attributes. For this specific example; remove all outfits with an above-zero "hyperdrive" attribute.

> Or you could remove all cooling, things like that

Similar when using attributes; remove all Outfits that have an above-zero "cooling" attribute.

The attributes seem to be able to already act as tags, so I don't see a benefit in adding a separate tag mechanism here.

I also agree with warp-core; just taking Outfits blindly based on some properties seems risky; what if some outfit provides both hyperdrive and engines; we actually do have [such an outfit](https://github.com/endless-sky/endless-sky/blob/83e99e8e01e5e1d7ecfa095135a1b7ec9176e667/data/sheragi/sheragi%20outfits.txt#L98) in-game.

**Comment by Zitchas** [comment]
I'm not sure about reviving based on attribute, but being able to do condition testing for branches, missions, etc. I see as being very useful.

**Comment by petervdmeer** [comment]
> I'm not sure about reviving based on attribute, but being able to do condition testing for branches, missions, etc. I see as being very useful.

Agreed, I also consider this very useful. But if we can achieve this useful functionality without having to add tags (by just re-using names and attributes information that we already have), then we can get those benefits with less maintenance effort on the Outfit creation. (And likely also more reliable, because names and attributes already need to be provided to make them functionally work, a forgotten tag only gets noticed somewhere later during storing writing.)

**Comment by Hurleveur** [comment]
What you describe would be powerful, but hard to write understandable syntax with so many possibilities, whereas the tag is already ready to go and pretty simple.
Tags would make sense to group outfits, so one can easily retrieve one "engine" without having to say it can either be jump drive or the same for all outfit classes, this makes total sense to me.

File: source/Outfit.h
Comment (warp-core) [inline source/Outfit.h:115]: This should be an `std::set`.

File: source/GameAction.cpp
Comment (warp-core) [inline source/GameAction.cpp:137]: What's going on here?
Why is there a UI object being passed here?
Why are you iterating over *every* outfit in the game rather htan just the outfits on the ship?

**Review by warp-core — COMMENTED** [review]
The event token should probably be something like `"take outfits with tags"` not just `outfittag`.

You should also provide a way to take outfits with any one of multiple tags or outfits with all of a set of tags. I would suggest doing this the way LocationFilter handles multiple attributes.

File: source/PlayerInfo.cpp
Comment (Hurleveur) [inline source/PlayerInfo.cpp:2703]: I think you should consider all ships and do different conditions like it was recently done by Amazinite
could be in a follow up PR though

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looks good I'm just a bit annoyed by the fact it can only happen on the flagship but that can be expanded later on.

File: source/PlayerInfo.cpp
Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2703]: hmmm... I will see

Comment (Hurleveur) [reply source/PlayerInfo.cpp:2703]: Basically a flagship/cargo/etc only option would be nice, but its best left for later prob

File: source/GameAction.cpp
Comment (petervdmeer) [inline source/GameAction.cpp:188]: I agree with all the use-cases given in the discussion, but I'm not on board yet with the keywords and concepts.

Not sure how to do this better, maybe by calling it meta-data instead of tags?   Outfits don't really get tagged with being "an engine", or being "human tech" or "republic tech", but a manufacturer or an intended use-case sounds a bit more like meta-data.

**Review by petervdmeer — CHANGES_REQUESTED** [review]

File: source/GameAction.cpp
Comment (RisingLeaf) [reply source/GameAction.cpp:188]: so i should just rename it?

Comment (petervdmeer) [reply source/GameAction.cpp:188]: Also, if we have functionality to remove outfits, then I expect such functionality to work on this "metadata", but also on attributes, so the syntax should be something like:
```
remove outfits
	ships
		<query to select the ships affected>
	outfits
		<query to select the affected outfits, including options to use attributes and this metadata>
```

Comment (Hecter94) [reply source/GameAction.cpp:188]: Why attributes? Can you give an example of why a content creator would want to remove something based on an attribute rather than by "metadata"? 

Bearing in mind that if an outfit/ship is missing valid metadata that a content creator needs, they can add that metadata in the same PR that requires the use of it.

Comment (petervdmeer) [reply source/GameAction.cpp:188]: Sure; A plugin creator that creates a mission where all Jump-drive Outfits get removed for some mission where the player can fly in-system to a nearby planet where he/she can buy a new jump-drive (The government of holy planet X confiscated all jump-drive capable equipment; you were warned not to land on this holy planet with active jump-drives. The moon near planet Y seems to sell expensive jump-drives which you can purchase if you wish to leave the system later).
The plugin creator cannot update the original outfit, and we already have attributes that provide us the metadata to determine which outfits contain jump-drive functionality, so no need to add this metadata once more.

Comment (RisingLeaf) [reply source/GameAction.cpp:188]: will look over it tomorrow

**Comment by RisingLeaf** [comment]
Ok, I just did a complete overhaul of the mechanic, I will update the description right now, therefore I will rerequest a review from everybody.

**Comment by RisingLeaf** [comment]
This PR will cooperate with #7515

**Review by petervdmeer — COMMENTED** [review]
As discussed on Discord;
This new syntax looks much better (much more generic). But we might want to go for LocationFilters and get an even stronger mechanism (with less maintenance).

**Comment by bene-dictator** [comment]
If I'm not misunderstanding... would it be possible to use these to reorder outfits in the outfitter? Right now it is all in alphabetical order, but this means that some related outfits are separated from others, like reactors (armageddon core, & radiothermal being on opposite ends of the Power section, for example), plasma thrusters and afterburners etc., or different alien technologies within the same outfitter (Remnant outfitters or the human outfitter on Greenwater, for example). Order by type/class of outfit first, then by alien type, then by alphabetical order.

**Comment by RisingLeaf** [comment]
> 

Not really the goal o this PR but maybe a usable side effect

**Comment by RisingLeaf** [comment]
> As discussed on Discord; This new syntax looks much better (much more generic). But we might want to go for LocationFilters and get an even stronger mechanism (with less maintenance).

I dont know if the latest changes are what you wanted, but I created an OutfitFilter class

File: source/Outfit.h
Comment (RisingLeaf) [inline source/Outfit.h:117]:
```suggestion
	std::set<std::string> tags;
```

File: source/PlayerInfo.cpp
Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2770]: I would like to leave this to a later PR

File: source/OutfitFilter.cpp
Comment (Hurleveur) [inline source/OutfitFilter.cpp:50]: I think you should add smth when the keyword isnt recognized (here, and in the not if)

File: EndlessSky.xcodeproj/project.pbxproj
Comment (Hurleveur) [inline EndlessSky.xcodeproj/project.pbxproj:30]: is this normal? Was it added twice? I mean the test runs I guess but feels weird to me

**Review by Hurleveur — CHANGES_REQUESTED** [review]
looks good, there are many things we can do with this

File: EndlessSky.xcodeproj/project.pbxproj
Comment (RisingLeaf) [reply EndlessSky.xcodeproj/project.pbxproj:30]: i dont really know, xcode will be gone soon anyways

File: source/OutfitFilter.cpp
Comment (RisingLeaf) [reply source/OutfitFilter.cpp:50]: ok

File: EndlessSky.xcodeproj/project.pbxproj
Comment (petervdmeer) [reply EndlessSky.xcodeproj/project.pbxproj:30]: The patterning-code is the only part of formation flying that was merged so far, so I guess this file is just not referenced from other files yet (in which case the game can compile fine without this file).

File: source/PlayerInfo.cpp
Comment (petervdmeer) [inline source/PlayerInfo.cpp:2782]: I voiced [some concerns](https://github.com/endless-sky/endless-sky/pull/7482#discussion_r1014729762) on using `tag` in the keyword; for me tagging is something that happens in-game, when something happens. This is data that is always, by default, on the Outfits, so I would really like to see something else here.

**Review by petervdmeer — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.cpp
Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2782]: any suggestion?
attribute would be ambiguos

File: EndlessSky.xcodeproj/project.pbxproj
Comment (RisingLeaf) [reply EndlessSky.xcodeproj/project.pbxproj:30]: xcode is removed now anyways

File: source/OutfitFilter.cpp
Comment (RisingLeaf) [reply source/OutfitFilter.cpp:54]: dam dam dam daaaa:
done

File: source/PlayerInfo.cpp
Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2835]: how about:
subcategory
identifier

**Comment by RisingLeaf** [comment]
Waiting on answer from peter, therefore removing waiting on OP

**Comment by Hurleveur** [comment]
Can you add an integration test for this? If you want I can help do that btw
Or at the very least test it manually to some extend, with screenshots, but I'd feel safer with integ test

**Comment by RisingLeaf** [comment]
> Can you add an integration test for this? If you want I can help do that btw
> 
> Or at the very least test it manually to some extend, with screenshots, but I'd feel safer with integ test

It would be nice if you could help me with the tests

**Comment by Hurleveur** [comment]
Cant you just have a test for: displaying how many marked outfits the ship has (whilst giving some of those outfits to the ship), then taking them away and counting them again to check they're not there anymore.

File: source/PlayerInfo.cpp
Comment (petervdmeer) [reply source/PlayerInfo.cpp:2835]: Maybe property or characteristic?

Comment (petervdmeer) [reply source/PlayerInfo.cpp:2835]: subcategory doesn't feel right, because this new keyword does not really identify categories (and Outfits are already part of a category). Identifier also doesn't feel right, because this new data does not uniquely identify a single outfit.

Comment (petervdmeer) [reply source/PlayerInfo.cpp:2835]: Thought about it a bit more; we could also just use tags here (as is currently done) and have the in-game functionality use another keyword. (There is no absolute requirement to use the keyword `tag` there.)

**Review by petervdmeer — COMMENTED** [review]
Neutral review: I didn't review the code yet, but also don't want to block this PR on my availability, so if someone else does a very good in-depth review then I'm ok with this PR going forward.

File: source/PlayerInfo.cpp
Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2835]: what do you mean with the "in-game functionality"?

Comment (petervdmeer) [reply source/PlayerInfo.cpp:2835]: Sorry, I wasn't clear. What I was thinking about is [a proposal to allow tagging/marking ships during missions](https://github.com/endless-sky/endless-sky/issues/4930).
It's a feature which I really want to have as a content-creator, because it will make it possible to also do some story-lines around ships.

But it is not very relevant; the feature in 4930 can also use another keyword, and then we can just use the keyword tag here.

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:2835]: If you are fine with it...
I mean I dont mind just keeping it how it is

**Pull request closed without merging**
